api/pkg/transformer/feast/bigtablestore/encoding.go

```diff
@@ -0,0 +1,425 @@
+package bigtablestore
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/bigtable"
+
+	feast "github.com/feast-dev/feast/sdk/go"
+	"github.com/feast-dev/feast/sdk/go/protos/feast/serving"
+	"github.com/feast-dev/feast/sdk/go/protos/feast/types"
+	"github.com/gojek/merlin/pkg/transformer/spec"
+	"github.com/golang/protobuf/proto"
+	"github.com/linkedin/goavro/v2"
+)
+
+// Encoder is used to encode OnlineFeaturesRequest into RowQuery
+type Encoder struct {
+	registry        CodecRegistry
+	featureSpecs    map[featureTableKey]*spec.FeatureTable
+	featureMetadata map[featureTableKey]*spec.FeatureTableMetadata
+}
+
+type featureTableKey struct {
+	project string
+	table   string
+}
+
+func entityKeysToBigTable(project string, entityKeys []*spec.Entity) string {
+	keyNames := make([]string, len(entityKeys))
+	for i, e := range entityKeys {
+		keyNames[i] = e.Name
+	}
+	return project + "__" + strings.Join(keyNames, "__")
+}
+
+func feastValueToStringRepr(val *types.Value) (string, error) {
+	switch v := val.Val.(type) {
+	case *types.Value_StringVal:
+		return val.GetStringVal(), nil
+	case *types.Value_Int64Val:
+		return strconv.FormatInt(val.GetInt64Val(), 10), nil
+	case *types.Value_Int32Val:
+		return strconv.FormatInt(int64(val.GetInt32Val()), 10), nil
+	case *types.Value_BytesVal:
+		return string(val.GetBytesVal()), nil
+	default:
+		return "", fmt.Errorf("unsupported value type %v", v)
+	}
+}
+
+func feastRowToBigTableKey(entity feast.Row, entityKeys []*spec.Entity) (string, error) {
+	entityValues := make([]string, len(entityKeys))
+	for i, key := range entityKeys {
+		entityValue, err := feastValueToStringRepr(entity[key.Name])
+		if err != nil {
+			return "", err
+		}
+		entityValues[i] = entityValue
+	}
+	return strings.Join(entityValues, "#"), nil
+}
+
+func compareEntityKeys(entityKeys1 []*spec.Entity, entityKeys2 []*spec.Entity) bool {
+	if len(entityKeys1) != len(entityKeys2) {
+		return false
+	}
+	for i := 0; i < len(entityKeys1); i++ {
+		if !proto.Equal(entityKeys1[i], entityKeys2[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+func getFeatureType(featureSpec *spec.FeatureTable, featureName string) (string, error) {
+	for _, f := range featureSpec.Features {
+		if f.Name == featureName {
+			return f.ValueType, nil
+		}
+	}
+	return "", fmt.Errorf("feature %s is not part of specification", featureName)
+}
+
+func avroToValueConversion(avroValue interface{}, featureType string) (*types.Value, error) {
```

Comment:
Can we have separate test case for `avroToValueConversion` currently it is only being tested for `STRING` case and I think this is quite critical part of the implementation